Adding documentation for certificate renewal (#2293)
dapr/docs: daprdocs/content/en/reference/cli/dapr-mtls/dapr-mtls-renew-certificate.md

---
type: docs
title: "mtls renew certificate CLI command reference"
linkTitle: "mtls renew certificate"
description: "Detailed information on the mtls renew certificate CLI command"
weight: 3000
---

### Description

> This command can be used to renew expiring Dapr certificates in Kubernetes cluster.

Reviewer: ```suggestion
This command can be used to renew expiring Dapr certificates. For example the Dapr Sentry service can generate default root and issuer certificates used by applications. For more information see [secure Dapr to Dapr communication]({{< ref secure-dapr-to-dapr-communication >}})
```

It renews root CA certificate, issuer certificate and issuer key.

### Supported platforms

- [Kubernetes]({{< ref kubernetes >}})

### Usage

```bash
dapr mtls renew-certificate [flags]
```

### Flags

| Name           | Environment Variable | Default           | Description                                 |
| -------------- | -------------------- | ----------------- | ------------------------------------------- |
| `--help`, `-h` |                      |                   | help for renew-certificate

> | `--kubernetes`, `-k` |                      | `false` | supprted platform|                             |

Reviewer: Ah! Perfect. You added the -k command for this! Excellent. Now we just need to add these for the expiry and export commands

| `--valid-until`  |                      | 365 days |  Validity for newly created certificates |
| `--restart`  |                      | false |  Restarts Dapr control plane services (Sentry service, Operator service and Placement server) |

> | `--timeout`  |                      | 300 sec |  The timeout for the certificate renewal process |

Reviewer: I do not understand this timeout. This is the time for the renewal process to complete? What if this is a 300 node K8s cluster, will this succeed?

Author: This is actually default Helm wait timeout for any process like install/upgrade to complete. If any operation is supposed to take more time, we can set this flag to higher value to stop timing out helm operation. see `helm install/upgrade -h`

> | `--ca-root-certificate`  |                      |  |  User provided root certificate pem file path|

Reviewer: ```suggestion
| `--ca-root-certificate`  |                      |  |  File path to user provided PEM root certificate
```

> | `--issuer-public-certificate`  |                      |  |  User provided issuer certificate pem file path|

Reviewer: ```suggestion
| `--issuer-public-certificate`  |                      |  |  File path to user provided PEM issuer certificate |
```

> | `--issuer-private-key`  |                      |  |  User provided issue private key file path|

Reviewer: ```suggestion
| `--issuer-private-key`  |                      |  |  File path to user provided PEM issue private key |
```

| `--private-key`  |                      |  |  User provided root.key file which is used to generate root certificate|

### Examples

> #### Renew certificates by generating fresh new certificates

Reviewer: ```suggestion
#### Renew certificates by generating brand new certificates
```

> Generates new root and issuer certificates for Dapr kubernetes cluster with a default validity of 365 days.

Reviewer: ```suggestion
Generates new root and issuer certificates for the Kubernetes cluster with a default validity of 365 days. The certificates are not applied to the Dapr control plane.
```

```bash
dapr mtls renew-certificate -k
```

> Generates new root and issuer certificates for kubernetes cluster with a default validity of 365 days and restart the control plane services.

Reviewer: ```suggestion
Generates new root and issuer certificates for the Kubernetes cluster with a default validity of 365 days and restarts the Dapr control plane services.
```

```bash
dapr mtls renew-certificate -k --restart
```

> Generates new root and issuer certificates for kubernetes cluster with a given validity.

Reviewer: ```suggestion
Generates new root and issuer certificates for the Kubernetes cluster with a given validity time.
```

```bash
dapr mtls renew-certificate -k --valid-until <no of days>
```

> Generates new root and issuer certificates for kubernetes cluster with a given validity and restart the control place services.

Reviewer: ```suggestion
Generates new root and issuer certificates for the Kubernetes cluster with a given validity time and restarts the Dapr control place services.
```

```bash
dapr mtls renew-certificate -k --valid-until <no of days> --restart
```
#### Renew certificate by using user provided certificates

> Rotates certificate of your kubernetes cluster with provided ca.pem, issuer.pem and issuer.key file path and restart the control plane services

Reviewer: ```suggestion
Rotates certificates for the Kubernetes cluster with the provided ca.pem, issuer.pem and issuer.key file paths and restarts the Dapr control plane services
```

```bash
dapr mtls renew-certificate -k --ca-root-certificate <ca.pem> --issuer-private-key <issuer.key> --issuer-public-certificate <issuer.pem> --restart
```

> Rotates certificate of your kubernetes cluster with provided ca.pem, issuer.pem and issuer.key file path.

Reviewer: ```suggestion
Rotates certificates for the Kubernetes cluster with the provided ca.pem, issuer.pem and issuer.key file paths.
```

```bash
dapr mtls renew-certificate -k --ca-root-certificate <ca.pem> --issuer-private-key <issuer.key> --issuer-public-certificate <issuer.pem>
```

> #### Renew certificates by generating fresh certificates using provided root private key

Reviewer: ```suggestion
#### Renew certificates by generating brand new certificates using the provided root private key
```

> Uses existing private root.key to generate new root and issuer certificates for kubernetes cluster with a given validity for created certs.

Reviewer: ```suggestion
Uses existing private root.key to generate new root and issuer certificates for the Kubernetes cluster with a given validity time for created certs.
```

Reviewer: What this topic is lacking is any of the guidance on best practices. So next we need to update the guidance in this document https://docs.dapr.io/operations/security/mtls/ especially in the https://docs.dapr.io/operations/security/mtls/#updating-root-or-issuer-certs section. This section was recently updated before these commands were available. I suggest that you talk with Bernd who was the person responsible for testing this out. 

If effect we want to be clear that the user has two primary choices on guidance showing how to put all the mtls commands together and these choice are;
1) Use the self-signed Dapr generated root and issuer certs which is easy, but you will encountered downtime on renewal
2) Use this renew command with a private root key to update the deploy and renew root and issuer certs to avoid downtime. We should show the complete set of steps and commands to accomplish this. 

Bottom line, please work with me and Bernd to update this topic https://docs.dapr.io/operations/security/mtls/ and in particular this section https://docs.dapr.io/operations/security/mtls/#updating-root-or-issuer-certs. For example lets split this topic into Kubernetes and self hosted.

Author: @msfussell  As of now these command's implementation is done only for k8s mode. So in the k8s section of https://docs.dapr.io/operations/security/mtls/#updating-root-or-issuer-certs we can update the best practice along with cli command available for that.
I think we can create a separate PR for updating this ? Kindly share your thoughts.
@berndverst

Reviewer: I think we should do this at the same time since it has to be done for v1.7 release anyway. The article above was updated before this new command existed and now we should ensure that the best guidance is explained. These CLI docs do not provide any guidance.

Author: @msfussell Update at the above link, The cli commands to be used for each scenario. Kindly share your thoughts on this.

```bash
dapr mtls renew-certificate -k --private-key myprivatekey.key --valid-until <no of days>
```

> Uses existing private root.key to generate new root and issuer certificates for kubernetes cluster with a default validity of 365 days for created certs.

Reviewer: ```suggestion
Uses the existing private root.key to generate new root and issuer certificates for the Kubernetes cluster.
```
